rfcs/0000-windows-containers.md

```diff
@@ -0,0 +1,132 @@
+## Problem:
+
+kpack does not support building containers that can run on windows
+
+## Outcomes:
+
+### kpack should run on clusters that have windows workers
+
+kpack service pods and build pods for linux images will need to know not to run on the windows workers to work in a mixed-os cluster.
+
+The kpack release configuration should specify to run pods explicitly on linux os.
+
+The kpack controller should schedule linux builds on linux workers.
+
+**Actions to Take**
+
+Use the following node selector for all service pods and build pods for linux images:
+
+```
+nodeSelector:
+    kubernetes.io/os: linux
+```
+
+### kpack should restrict cluster stacks to be os-specific
+
+This would improve simplicity and usability. If you create mixed-os resources, it may be difficult to get out of the situation.
+
+If a user creates a cluster stack with build and run images of mixed os, the resource should be marked as not-ready
+
+**Actions to Take**
+
+- Add this validations to the kpack-controller
+
+Notes:
+- Stores will not be aware of the buildpackages os. There will be no constraint on creating stores with windows buildpackages and linux buildpackages.
+
+### kpack should support creating windows builders
+
+This is currently supported in the [pack cli](https://github.com/buildpacks/pack/issues/469)
+
+**Actions to Take**
+
+- kpack-controller will use a new environment variable `WINDOWS_LIFECYCLE_IMAGE` to reference the windows lifecycle image
+- Provide the kpack release (`release.yaml`) with a windows version of the lifecycle image and configure it with `WINDOWS_LIFECYCLE_IMAGE`
+- When a Builder or ClusterBuilder is created using a windows stack, kpack-controller must create a windows builder using the windows lifecycle image.
+    - Image os is determined by the image [os property](https://github.com/opencontainers/image-spec/blob/master/config.md#:~:text=os%20string).
+
+Notes:
+
+- Builders will not be aware of the buildpackages os. The only compatibility constraint between stacks and buildpackages for creating builders will be the stack id.
+
+### kpack should support running builds on windows workers
+
+**Actions to Take**
+
+- kpack-controller will accept new configuration via environment variable(s) for windows image(s) needed to run builds. The current behavior that must be recreated on windows is contained in the following images:
+    - `BUILD_INIT_IMAGE` - getting the build ready with creds and source
+    - `COMPLETION_IMAGE` - an image to run as a required regular (non-init) container
+    - `REBASE_IMAGE` - run a rebase
+    - We'll call these "helper images"
+    - Out of scope:
+        - What to name the windows versions of these images
```

Comment:
> What to name the windows versions of these images

Not about the naming... but more about do we actually need windows specific versions for ALL of these images? Unless intentionally choosing to go down that path.
What particularly stood out for me was the `COMPLETION_IMAGE ` (and maybe the `BUILD_INIT_IMAGE` to a certain degree - not so sure how it works).

The `COMPLETION_IMAGE`, if I recollect correctly, all it did was show a completion message. But more importantly, there was a recent discussion about using this image for Image signing. And I am not clear if we need a linux and windows variant for that.

Reply:
the completion image also serves the requirement of having at least one regular (non-init) container which will need to run on the windows node